rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,479 @@
+    key_usage: RsaKeyUsage,
+    num_p_exp: u32, // !REVIEW this is not needed since the public exponent will always be one
+}
+
+impl fmt::Debug for RsaKeyPair {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RsaKey")
+            .field(
+                "inner",
+                &format_args!("{:p}", *self.inner.as_ref().get_ref()),
+            )
+            .field("key_mode", &self.key_mode)
+            .field("key_usage", &self.key_usage)
+            .finish()
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `RsaKeyMode` will indicate if this [`RsaKeyPair`] has [`RsaKeyMode::PublicOnly`] key or a [`RsaKeyMode::KeyPair`].
+pub enum RsaKeyMode {
+    PublicOnly,
+    KeyPair,
+}
+
+impl RsaKeyMode {
+    fn to_u32(&self) -> symcrypt_sys::UINT32 {
+        match self {
+            RsaKeyMode::PublicOnly => 0 as symcrypt_sys::UINT32,
+            RsaKeyMode::KeyPair => 2 as symcrypt_sys::UINT32,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+/// `RsaKeyUsage` will indicate if this [`RsaKeyPair`] will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`].
+pub enum RsaKeyUsage {
+    Sign,
+    Encrypt,
+}
+
+// !REVIEW: Based on the discussions i've had with Phil, and Sam offline, im not sure if we want to include the minimal validation or no fips flags as the purpose of this
+// code will be to eventually be fips certified. We can include if the group thinks otherwise though. 
+
+// #define SYMCRYPT_FLAG_KEY_NO_FIPS               (0x100)
+// #define SYMCRYPT_FLAG_KEY_MINIMAL_VALIDATION    (0x200)
+// #define SYMCRYPT_FLAG_RSAKEY_SIGN       (0x1000)
+// #define SYMCRYPT_FLAG_RSAKEY_ENCRYPT    (0x2000)
+impl RsaKeyUsage {
+    pub fn to_flag(&self) -> symcrypt_sys::UINT32 {
+        // making public so callers can access underlying flag
+        match self {
+            RsaKeyUsage::Sign => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN,
+            RsaKeyUsage::Encrypt => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT,
+        }
+    }
+}
+
+pub enum PrimeIndex {
+    First,
+    Second,
+}
+
+impl PrimeIndex {
+    pub fn to_u32(self) -> u32 {
+        match self {
+            PrimeIndex::First => 0,
+            PrimeIndex::Second => 1,
+        }
+    }
+}
+
+/// !REVIEW: I've tried to closely follow SymCrypts API calls and semantics making some simplifications, but I think we have the opperuntiy to simplify further if we chose.
+/// My questions / Suggestions
+/// 1. We've introduced the idea of the RsaKeyPair state, which can hold a lot of data, and makes a lot of the calls to the underlying SymCrypt code redundant. most of the accessors no longer
+/// need to call SymCrypt since the same data is stored on the state.
+/// 2. The idea of new() then generate_key() / generate_set_public_only() / set_key_pair() is a bit convoluted IMO, we can squash these together to something like: 
+/// new_random(), and  new_set_public() and new_set_key_pair().
```

Comment:
Seems like after the RsaKeyPair object is created. It should be ready to do a sign/verify. It shouldn't change.

So the generate or import should return a new RsaKeyPair.